doc/thingpedia-device-class.md

```diff
@@ -1,2 +1,113 @@
+that require no authentication or those only require an API key.
+More options such as OAuth and IoT discovery are described in the
+[complete guide for authentication and discovery](/doc/thingpedia-device-intro-auth-n-discovery.md).
+
+## Function declarations
+The user interacts with your Thingpedia device through two types of functions: queries and actions. 
+Thus, the first step is to decide on what queries and actions you should expose.
+
+The only requirement imposed from Thingpedia is that queries are free of side-effects and can return results (as output parameters),
+and actions have side effects, and cannot return results. Other than that, the design of which functions to include is highly device specific.
+
+### Qualifiers
+Additional qualifier can be used to specify the property of the query. 
+A query is `monitorable` if it's meaningful to monitor its return value and get triggered. 
+A query is a `list` query if it normally returns multiple results. 
+
+For example, a query to return latest emails would be considered both `monitorable` and `list`,
+and its declaration will look like:
+```tt
+monitorable list latest_emails (...); 
+```
+
+For The Cat API, it returns a random cat every time, so it's not reasonable to monitor it. 
+It also returns a single cat by default, so we also don't mark it as `list`.
+
+### Parameters
+To take full advantage of the functionality we provided in ThingTalk (filtering, chaining, etc.),  
+every parameter needed for ___both input and output___ should be listed. 
+A parameter is described as follows: 
+```tt
+[in req | in opt | out] <name> : <type>
+```  
+
+A parameter can be either `in req` (required input), `in opt` (optional input), or `out` (output).
+The `type` of a parameter could be: String, Number, Boolean, Date, Time, Location, 
+Entity(...), Enum(...), Measure(...), etc.
+For the full list, see the [ThingTalk reference](/doc/thingtalk-reference.md)
+
+For The Cat API, we have only a query to return random cat pictures. 
+It takes an optional input parameter `count` of type Number, 
+and 3 output parameters of type Entities: `image_id`, `picture_url`, and `link`.
+
+### Annotations
+Annotations are used to provide additional information to the corresponding code. 
+Annotations always come after the code and before the semicolon. 
+There are two types of annotations: natural language annotation and implementation annotation.
+
+#### Natural language annotation
+Natural language annotation, as its name, is the annotation related to natural language 
+and it will be translated to different languages based on users' profile. 
+It's  denoted by `#_[<key>=<value>]`.
+
+Here is a list of required natural language annotations for functions:
+- `canonical`: The canonical form of the function name, used by the semantic parser (certain versions);
+it's a good idea to omit stop words for this, and to use a longer expression. 
+You must omit all arguments from the canonical form.
```

Comment:
By the way, canonical is optional. If omitted, it defaults to the `clean`-ed function name + "on" + `clean`-ed device name.
And it would be best to describe it as the "human readable" function name - because that's what it is. It's primarily used by Brassau these days.